### PR TITLE
Added zabbix_action module 

### PR DESCRIPTION
**Pull request by rubentsirunyan**
##### SUMMARY

Implements zabbix_action module to manipulate Zabbix action objects on Zabbix server via API.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

zabbix_action

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (resmo) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py]: I would use the key of the common return `msg` instead of `result`, as it is basically a message

https://docs.ansible.com/ansible/2.7/reference_appendices/common_return_values.html#msg

**Comment by resmo** [comment]
does not look too bad, I would label it shippable, @D3DeFi any comments?

(it actually looks very good, good job!)

**Comment by D3DeFi** [comment]
Agree @resmo , very nice and readable addition. I was scrolling through it earlier today and wanted to test it with multiple zabbix-server versions, but haven't had that much time to actually do it. I can test it tomorrow or we can ship it right away if you want.

@rubentsirunyan thank you for your awesome contribution!

**Comment by resmo** [comment]
waiting for your feedback after testing

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (rubentsirunyan) [reply lib/ansible/modules/monitoring/zabbix/zabbix_action.py]: Agree. Will replace it today.

Comment (rubentsirunyan) [reply lib/ansible/modules/monitoring/zabbix/zabbix_action.py]: Done.

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py]: looks like string representation of time (e.g. `1h`) was introduced in Zabbix-server 3.4, before that this option used integer from range (60..XY) - https://www.zabbix.com/documentation/3.2/manual/api/reference/action/object

Running it like this allowed me to create action on 3.0 as well:
```yaml
          zabbix_action:
            server_url: '{{ zbx_server_url }}'
            login_user: '{{ zbx_user }}'
            login_password: '{{ zbx_password }}'
            validate_certs: False
            name: "Send alerts to Admin"
            event_source: 'trigger'
            esc_period: 60  # added this
            state: present
            status: enabled
            conditions:
              - type: 'trigger_severity'
                operator: '>='
                value: 'Information'
            operations:
              - type: send_message
                subject: "Something bad is happening"
                message: "Come on, guys do something"
                media_type: 'Email'
                send_to_users:
                  - 'Admin'
```
When left as default I've got following error with 3.0 release:
```
fatal: [zbx-dev-node002 -> localhost]: FAILED! => {"changed": false, "msg": "Failed to create action 'Send alerts to Admin': (u'Error -32602: Invalid params., Action \"Send alerts to Admin\" has incorrect value for \"esc_period\" (minimum 60 seconds). while sending {\"p$
rams\": {\"operations\": [{\"opmessage\": {\"mediatypeid\": \"1\", \"default_msg\": \"0\", \"message\": \"Come on, guys do something\", \"subject\": \"Something bad is happening\"}, \"opmessage_usr\": [{\"userid\": \"1\"}], \"operationtype\": \"0\"}], \"status\": \"0\",
\"esc_period\": \"1h\", \"r_shortdata\": null, \"r_longdata\": null, \"def_shortdata\": null, \"name\": \"Send alerts to Admin\", \"recovery_operations\": null, \"acknowledge_operations\": null, \"def_longdata\": null, \"filter\": {\"evaltype\": \"0\", \"conditions\": [{
\"operator\": \"5\", \"conditiontype\": \"4\", \"value\": \"1\"}]}, \"ack_shortdata\": null, \"eventsource\": \"0\", \"ack_longdata\": null}, \"jsonrpc\": \"2.0\", \"method\": \"action.create\", \"auth\": \"c5747ac07dbca5c565579925d4799678\", \"id\": 4}', -32602)"}
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py]: just a minor typo and nothing important, multiple occurrences of 'compaire'

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:1567]: [recovery_operations](https://www.zabbix.com/documentation/3.2/manual/api/reference/action/create) supported since zbx-server 3.2, ignored on 3.0, fails on lower versions

```
fatal: [zbx-dev-node001 -> localhost]: FAILED! => changed=false
  msg: 'Failed to create action ''Send alerts to Admin'': (u''Error -32602: Invalid params., Incorrect parameter for action "Send alerts to Admin". while sending {"params": {"operations": [{"opmessage": {"mediatypeid": "1", "default_msg": "0", "message": "Come on, guys d
o something", "subject": "Something bad is happening"}, "opmessage_usr": [{"userid": "1"}], "operationtype": "0"}], "status": "0", "esc_period": "60", "r_shortdata": null, "r_longdata": null, "def_shortdata": null, "name": "Send alerts to Admin", "recovery_operations": n
ull, "acknowledge_operations": null, "def_longdata": null, "filter": {"evaltype": "0", "conditions": [{"operator": "5", "conditiontype": "4", "value": "1"}]}, "ack_shortdata": null, "eventsource": "0", "ack_longdata": null}, "jsonrpc": "2.0", "method": "action.create", "
auth": "1a63b4bb1d45604a5aec26aa43651adb", "id": 4}'', -32602)'
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:1568]: [acknowledge_operations](https://www.zabbix.com/documentation/3.4/manual/api/reference/action/create) supported since zbx-server 3.4, ignored in 3.2/3.0, fails in lower versions

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:746]: both [ack_longdata and ack_shortdata](https://www.zabbix.com/documentation/3.4/manual/api/reference/action/object#action) were introduced in zbx-server 3.4, adding them to API call doesn't break calls to 3.2 and 3.0 versions tho, < 3.0 versions are failing

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:350]: Pasting whole example here so you can better understand me.

using recovery and acknowledge operations:
* is not idempotent on zbx-server 3.4 and 4.0 (result always `changed=True`)
* fails on 3.0 and lower releases (probably related to my comments above, that this was supported later)
* Action configured with recovery and acknowledge operations cannot have these operations removed by update, this is possible only by deleting action first and recreating it again.

```yaml
- block:
    - name: Deploy trigger action (new)
      zabbix_action:
        server_url: '{{ zbx_server_url }}'
        login_user: '{{ zbx_user }}'
        login_password: '{{ zbx_password }}'
        validate_certs: False
        name: "Send alerts to Admin"
        event_source: 'trigger'
        state: present
        status: enabled
        conditions:
          - type: 'trigger_severity'
            operator: '>='
            value: 'Information'
        operations:
          - type: send_message
            subject: "Something bad is happening"
            message: "Come on, guys do something"
            media_type: 'Email'
            send_to_users:
              - 'Admin'
        recovery_operations:
          - type: send_message
            subject: "Host is down"
            message: "Come on, guys do something"
            media_type: 'Email'
            send_to_users:
              - 'Admin'
        acknowledge_operations:
          - type: send_message
            media_type: 'Email'
            send_to_users:
              - 'Admin'
      register: result

    - assert:
        that:
          - "result.changed"
      ignore_errors: True

    - name: Deploy trigger action (existing)
      zabbix_action:
        server_url: '{{ zbx_server_url }}'
        login_user: '{{ zbx_user }}'
        login_password: '{{ zbx_password }}'
        validate_certs: False
        name: "Send alerts to Admin"
        event_source: 'trigger'
        state: present
        status: enabled
        conditions:
          - type: 'trigger_severity'
            operator: '>='
            value: 'Information'
        operations:
          - type: send_message
            subject: "Something bad is happening"
            message: "Come on, guys do something"
            media_type: 'Email'
            send_to_users:
              - 'Admin'
        recovery_operations:
          - type: send_message
            subject: "Host is down"
            message: "Come on, guys do something"
            media_type: 'Email'
            send_to_users:
              - 'Admin'
        acknowledge_operations:
          - type: send_message
            media_type: 'Email'
            send_to_users:
              - 'Admin'
      register: result

    - assert:
        that:
          - "not result.changed"
      ignore_errors: True

    - name: Deploy trigger action (remove rec&ack)
      zabbix_action:
        server_url: '{{ zbx_server_url }}'
        login_user: '{{ zbx_user }}'
        login_password: '{{ zbx_password }}'
        validate_certs: False
        name: "Send alerts to Admin"
        event_source: 'trigger'
        state: present
        status: enabled
        conditions:
          - type: 'trigger_severity'
            operator: '>='
            value: 'Information'
        operations:
          - type: send_message
            subject: "Something bad is happening"
            message: "Come on, guys do something"
            media_type: 'Email'
            send_to_users:
              - 'Admin'
      register: result

    - assert:
        that:
          - "result.changed"
      ignore_errors: True

  delegate_to: localhost
```

```
TASK [Deploy trigger action (new)] ********************************************************************************************************************************************************************************************************************************************
changed: [zbx-dev-node002 -> localhost]
changed: [zbx-dev-node003 -> localhost]

TASK [assert] *****************************************************************************************************************************************************************************************************************************************************************
ok: [zbx-dev-node002 -> localhost] => changed=false
  msg: All assertions passed
ok: [zbx-dev-node003 -> localhost] => changed=false
  msg: All assertions passed

TASK [Deploy trigger action (existing)] ***************************************************************************************************************************************************************************************************************************************
changed: [zbx-dev-node002 -> localhost]
changed: [zbx-dev-node003 -> localhost]

TASK [assert] *****************************************************************************************************************************************************************************************************************************************************************
fatal: [zbx-dev-node002 -> localhost]: FAILED! => changed=false
  assertion: not result.changed
  evaluated_to: false
  msg: Assertion failed
...ignoring
fatal: [zbx-dev-node003 -> localhost]: FAILED! => changed=false
  assertion: not result.changed
  evaluated_to: false
  msg: Assertion failed
...ignoring

TASK [Deploy trigger action (remove rec&ack)] *********************************************************************************************************************************************************************************************************************************
ok: [zbx-dev-node002 -> localhost]
ok: [zbx-dev-node003 -> localhost]

TASK [assert] *****************************************************************************************************************************************************************************************************************************************************************
fatal: [zbx-dev-node002 -> localhost]: FAILED! => changed=false
  assertion: result.changed
  evaluated_to: false
  msg: Assertion failed
...ignoring
fatal: [zbx-dev-node003 -> localhost]: FAILED! => changed=false
  assertion: result.changed
  evaluated_to: false
  msg: Assertion failed
...ignoring
```

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:279]: Add "- Works only with >= Zabbix 3.2" same probably goes for other arguments related to `recovery_*`

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_action.py:285]: Add "- Works only with >= Zabbix 3.4" same probably goes for other arguments related to `acknowledge_*`

**Review by D3DeFi — COMMENTED** [review]
I am usually testing changes against all current Zabbix-server LTS releases [2.2, 3.0, 4.0](https://www.zabbix.com/life_cycle_and_release_policy). There is, however, no rule established, which versions should be supported in ansible zabbix modules.

Note that I've basically just tested cases documented in `EXAMPLES`.

This module doesn't work with 2.2 release and requires explicitly setting `esc_period` to `60` instead of default `1h` to work on 3.0 release. (See my comments below).

Works as intended for 3.4 and 4.0, except a single idempotency problem I am mentioning later in my comments.

I would say there are these options:
* code support for older versions of zabbix (2.2 end of limited support - 31 August, 2019; 3.0 end of full support - 28 February, 2019)
* explicitly specify, that this module requires Zabbix-server 3.4 or higher (should go somewhere to `requirements`: in `DOCUMENTATION` variable? @resmo)
* change default values from string representations (e.g. `1h`) to integers (e.g. `60`) for options like `esc_period` and specify that this module requires Zabbix-server 3.0 or higher to run and Zabbix-server 3.4 to use ACK & recovery operations on actions (this should go to documentation under each module argument afaik - see my comments below)

IMO we can go with the 2nd option and lock this module only for 3.4+ releases. Idempotency fix can be delivered later via PR as I don't think that is something groundbreaking for now. What do you think guys?

@rubentsirunyan with merging of this PR you're becoming module maintainer so I am interested in your opinion in this as well :)

**Comment by D3DeFi** [comment]
Sorry for verbosity, I got carried away a bit

**Comment by rubentsirunyan** [comment]
@D3DeFi Thanks for the detailed review. I would go with the third option (+ delivering idempotency fix later via PR). I'll do the necessary changes and update the PR today.

File: lib/ansible/modules/monitoring/zabbix/zabbix_action.py
Comment (rubentsirunyan) [reply lib/ansible/modules/monitoring/zabbix/zabbix_action.py:350]: Found the idempotency issue. Zabbix API uses `recovery_operations` key when you **create** an action and `recoveryOperations` when you **get** an action. I need to do some transformation before comparing the data.

**Comment by rubentsirunyan** [comment]
@D3DeFi resolved all the issues including the idempotency issue.

**Review by D3DeFi — APPROVED** [review]

**Comment by D3DeFi** [comment]
Thank you for fixing it on such a short notice :)

**Comment by D3DeFi** [comment]
shipit

**Comment by resmo** [comment]
Merging... Thanks @rubentsirunyan, good job! And thanks @D3DeFi for the review.

**Pull request merged**
